### ISSUE TITLE
Use lower logging levels?

### ISSUE THREAD

**Issue by dpc** [n0-computer/iroh#3071]
I started playing with Iroh and I see it logs things on WARN etc. levels which by default is getting in a way of debugging my own code.

My philosophy of logging levels is that since they (somewhat unfortunately) are typically globally shared, they should be used with an end to end point of view.

I'd expect Iroh will always be used as a lower level infrastructure layer, and thus what's happening inside it will always be some kind of trace information, maybe debug for most important iroh logging statements, and not application level "information", "warning" or "error" which should be reserved for applications.

**Comment by flub** [comment]
Hi, in general we agree that not much should be logged on ERROR, WARN, and not even that much on INFO.  I think the rough guideline is that ERROR is a serious issue, potentially rendering the entire iroh `Endpoint` broken, while WARN is for things that are likely to be bugs.

So this makes my question to you: which message are you seeing specifically?  Maybe we have bugs to fix, or maybe we did get the log level wrong (there certainly are still those cases about).

Also in the interim, are you aware of the [EnvFilter directives](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#directives) which allow tweaking log levels per crate and module?

**Comment by dpc** [comment]
The core guidance IMO is that it all needs to be considered from the end user perspective.

If Iroh is hitting condition that means that the application is likely screwed and needs immediate user/administrator attention than it might deserve an `error` logging, sure, maybe. But it's rarely the case. An `Endpoint` simply being broken will rarely be a completely unexpected condition or something that the user can do anything about. The higher level application code using `Endpoint` will receive errors from `Endpoint` related methods anyway and then the app can log them as it see fits: if this `Endpoint` not working is actually a critical problem and unexpected, it will get logged as `error`.



That's another reason why libraries should not need to log all that much anyway - they should be returning errors to the app and let it handle it. Sometimes it is even worthwhile to expose something like a broadcast channel where the application can register to handle events (like errors) as it see fits. This is useful for exposing things from background threads etc. that can't simply return errors upwards to the higher level code.

BTW.  I wish `error` was renamed to `crit` because over and over devs conflate it, and mistakenly log every error, no matter how expected and mundane at `error` logging level, but that's another story.

> Also in the interim, are you aware of the [EnvFilter directives](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#directives) which allow tweaking log levels per crate and module?

Sure I am. I have spent quite a bit of time in Rust logging, e.g. being an author of `slog` logging ecosystem.

It is a workable bandaid, yes, but very crude. Typically applications end up maintaining [some long list of muted dependencies](https://github.com/fedimint/fedimint/blob/840f993805754c8ec7199db19a02fc33ffe8eb83/fedimint-logging/src/lib.rs#L112), and having dependency simply muted like that is less than ideal. Over last few years I've reported similar issues to many libraries. :)

**Comment by flub** [comment]
We do have some places where things can be problems, bugs either in use or -more common- implementation, where the problem is not fatal and we can not error out to the user.  The nature of all the things the Endpoint needs to do under the hood makes this unfortunately reasonably common.  In those cases we opt for error or warning logs so that we have a chance of getting bug reports etc.

One point were we might digress from your vision is that we are fairly verbose on debug level.  And this is generally essential for debugging user problems.  I expect as the library matures that these lower logging levels will slowly reduce.

I still think we essentially agree with the use of logging though.  So please, if you have specific log messages that are showing up too much do point them out so we can figure out if there are some bugs or whether we need to tweak the logging.

**Comment by dpc** [comment]
> One point were we might digress from your vision is that we are fairly verbose on debug level. And this is generally essential for debugging user problems. I expect as the library matures that these lower logging levels will slowly reduce.

Debug is a fair game. Ideally, it would be showing things relevant in debugging possible _application_'s issues related to Iroh, but ... debug is debug. All logging messages are fair in love and debugging or something.

**Comment by dpc** [comment]
```
2024-12-28T07:37:01.956378Z  INFO ep{me=zxagwbdyflz6ghm5}:magicsock:actor: iroh_net::magicsock: home is now relay https://use1-1.relay.iroh.network
./, was None
2024-12-28T07:37:01.956649Z  INFO ep{me=zxagwbdyflz6ghm5}:magicsock:relay-actor: iroh_net::magicsock::relay_actor: adding connection to relay: https://use1-1.relay.iroh.network./ for home-keep-alive
```

**Comment by flub** [comment]
So #3080 does address the 2nd one.  The first one I'm not convinced yet that I want to remove it.  Typically this will be emitted once at startup, which is very reasonable I think.  And it is a nice log message to stand out from all the debugging stuff.

**Comment by flub** [comment]
https://github.com/n0-computer/net-tools/pull/8 is also a bunch that I noticed and definitely don't want in the logs.  Maybe those ones were a regression.

**Comment by dpc** [comment]
> [n0-computer/net-tools#8](https://github.com/n0-computer/net-tools/pull/8) is also a bunch that I noticed and definitely don't want in the logs


Cool. I will eventually pick up the newest release and if I spot anything I will report back. Feel free to close or not the issue. I can keep posting anyway.

> Maybe those ones were a regression.

BTW. I recommend setting up `semgrep` or something like it to enforce things like that. In our project [we have bunch of things to enforce for various reasons](https://github.com/fedimint/fedimint/blob/master/.config/semgrep.yaml): wasm support, logging preferences, project-specific lints.

**Comment by flub** [comment]
`semgrep` looks interesting!  Thanks for the tip.

Will close this for now, but yes, feel free to call out other things.

**Comment by dpc** [comment]
We've noticed today that in real-like settings our logs party hard, with `disco` printing warnings all over the place (while everything seems to work).

**Comment by matheus23** [comment]
Yeah agreed.
A list of which ones exactly would be nice so we can attack them in one go with a PR.

I know of at least `send relay: message dropped, channel to actor is full` under saturated network traffic.

**Comment by dpc** [comment]
```
fedimintd-1  | 2025-03-19T18:44:01.446658Z  WARN ep{me=a555864de0}:magicsock:net_report.actor:reportgen.actor:probe: iroh_net_report::reportgen: pr
obe failed: Failed to send STUN request: StunIpv4: Invalid argument (os error 22) probe=StunIpv4 { delay: 263.995677ms, node: RelayNode { url: Rela
yUrl("https://euw1-1.relay.iroh.network./"), stun_only: false, stun_port: 3478, quic: Some(RelayQuicConfig { port: 7842 }) } }
fedimintd-1  | 2025-03-19T18:44:01.446667Z  WARN ep{me=a555864de0}:magicsock:net_report.actor:reportgen.actor:probe: iroh_net_report::reportgen: no
 successful probes in ProbeSet probe_proto=Some(StunIpv4)
```

These stand out.

**Comment by flub** [comment]
Oh, but these are not just bad logging.  These are things that need to be debugged.  Something strange is going on.

I think `Failed to send STUN request: StunIpv4: Invalid argument (os error 22)` is one issue.

It's hard to know right away whether ` UDP recv quic packets: no node state found` and `failed to open disco box err=Aead(Error)` are related or not right now.  There's a small chance they are, because something seems to be going very wrong.

Could you open new bugs for these with as much context as you can provide?

**Comment by dpc** [comment]
That's great to know. Might explain why our consensus was so slow (we were planning to debug it eventually). Created https://github.com/n0-computer/iroh/issues/3243

**Comment by dpc** [comment]
```
2026-03-18T03:48:45.816778Z  INFO endpoint{id=2f9002313f}:relay-actor: iroh::socket::transports::relay::actor: home is now relay https://usw1-1.relay.n0.iroh-canary.iroh.link./, was None
```

**Comment by flub** [comment]
> ```
> 2026-03-18T03:48:45.816778Z  INFO endpoint{id=2f9002313f}:relay-actor: iroh::socket::transports::relay::actor: home is now relay https://usw1-1.relay.n0.iroh-canary.iroh.link./, was None
> ```

This log should only appear once for each endpoint, unless you significantly move location or th home relay goes down and you need to switch to another one.

Are you finding this logs more often? Or do you find the described logging already too noisy? This one was a deliberate choice so far.

**Comment by dpc** [comment]
I think this should be a debug logging level. I'd consider info normal informative messages for the users of my app, and what relay is iroh using deep under the hood is way too much of an implementation detail for them already.

**Comment by flub** [comment]
> I think this should be a debug logging level. I'd consider info normal informative messages for the users of my app,

Whoa, that is a way more strict interpretation than usual I think. That would mean no library is ever allowed to log something on info or above. If you show info logs directly to your users I think you'd need to filter this to a specific tracing target that is designed for this.

**Comment by dpc** [comment]
> Whoa, that is a way more strict interpretation than usual I think. That would mean no library is ever allowed to log something on info or above.

Yeah, basically. See my first post in this issue. From my experience I'd say most libraries conform to this naturally.
